src/domain/dex/minimum_surplus.rs

```diff
@@ -0,0 +1,166 @@
+//! Minimum surplus requirements for DEX swaps.
+
+use {
+    crate::{
+        domain::{auction, dex::shared, eth},
+        util::conv,
+    },
+    bigdecimal::{BigDecimal, One, Zero},
+    ethereum_types::U256,
+    std::cmp,
+};
+
+/// DEX swap minimum surplus limits.
+#[derive(Clone, Debug)]
+pub struct MinimumSurplusLimits {
+    /// The relative minimum surplus (percent) required for swaps.
+    pub relative: BigDecimal,
+    /// The absolute minimum surplus required for swaps.
+    pub absolute: Option<eth::Ether>,
+}
+
+impl MinimumSurplusLimits {
+    /// Creates a new minimum surplus limits configuration.
+    pub fn new(relative: BigDecimal, absolute: Option<eth::Ether>) -> Result<Self, anyhow::Error> {
+        anyhow::ensure!(
+            relative >= BigDecimal::zero(),
+            "minimum surplus relative tolerance must be non-negative"
+        );
+        Ok(Self { relative, absolute })
+    }
+
+    /// Returns the minimum surplus for the specified token amount.
+    pub fn relative(&self, asset: &eth::Asset, tokens: &auction::Tokens) -> MinimumSurplus {
+        let absolute_as_relative = shared::absolute_to_relative(self.absolute, asset, tokens);
+
+        MinimumSurplus::new(cmp::max(
+            self.relative.clone(),
+            absolute_as_relative.unwrap_or(BigDecimal::zero()),
+        ))
+    }
+}
+
+/// A relative minimum surplus requirement.
+#[derive(Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord)]
+pub struct MinimumSurplus(BigDecimal);
+
+impl MinimumSurplus {
+    /// Creates a new minimum surplus from a decimal value.
+    pub fn new(value: BigDecimal) -> Self {
+        Self(value)
+    }
+
+    /// Creates a minimum surplus from an absolute amount and reference amount.
+    fn from_amount(surplus_amount: U256, reference_amount: U256) -> Self {
```

Comment:
`from_amount`, `apply`, and `as_factor` are not used at all.
`round` is only used in tests.
IMO these functions should be removed to not introduce code that is dead on arrival.